### ISSUE TITLE
declaration-block-no-ignored-properties

### ISSUE THREAD

**Issue by evilebottnawi** [stylelint/stylelint#906]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Even though you can define any group of properties together in a CSS rule, some of them will be ignored due to the display of the element.

As such, the rule warns when it finds:

- display: inline used with width, height, margin, margin-top, margin-bottom, and float.
- display: inline-block used with float.
- display: block used with vertical-align.
- display: table-* used with margin (and all variants) or float.

It is good rule from CSSComb. We should have the same. More info:
https://github.com/CSSLint/csslint/wiki/Require-properties-appropriate-for-display

> Which rule, if any, is this issue related to?

`property-display-no-effect-properties`

> What CSS is needed to reproduce this issue?

Example with display: https://github.com/CSSLint/csslint/wiki/Require-properties-appropriate-for-display

> What stylelint configuration is needed to reproduce this issue?

no matter

> Which version of stylelint are you using?

`5.0.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

no matter

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

no matter

> What did you expect to happen?

Throw errors.

> What actually happened (e.g. what warnings or errors you are getting)?

Nothing

**Comment by jeddy3** [comment]
Great suggestion!

Do you know if that is a definitive list of display values and ignored properties? There are [quite a few](https://developer.mozilla.org/en-US/docs/Web/CSS/display#Specifications) specs surrounding the `display` property, and it'd would be good to know we've got everything covered off.

For example, it looks like `display: flex` [also ignores floats](https://drafts.csswg.org/css-flexbox-1/#flex-items).

Do we make this generic and also check for other ignored properties. For example, `top`, `bottom`, `right` and `left` for [non-positioned](https://developer.mozilla.org/en/docs/Web/CSS/top) i.e. if `position: static` is set then flag if any of these four properties is also in the declaration block.

If we go down that route, then this rule is probably best called `declaration-block-no-ignored-properties`.

**Comment by evilebottnawi** [comment]
@jeddy3 be cool if we see other property and values, as is position and etc.

**Comment by davidtheclark** [comment]
I agree that this would be an awesome rule. I myself often make the mistake of putting margins on an inline element.

> For example, top, bottom, right and left for non-positioned i.e. if position: static is set then flag if any of these four properties is also in the declaration block.

Something different about this idea, @jeddy3, is that it's the inverse principle, right? Instead of checking that two properties *don't* occur together, it's saying that two properties *must* occur together. Both principles have the limitation that they can really only compare properties *within rules*, not across rules, even if in the browser different rules will end up applying to the same element. However, for the dont-occur-together principle, that's just a limitation: we can complain if `display: inline` and `margin: 10px` occur in the same rule, but we can't know if they're going to be applied via multiple rules to the same element. With the must-occur-together principle, it's not just a limitation: it's an assumption that will raise false positives. I might have good reasons for putting `position: absolute` in a different rule than my `top: 10px` declaration. Does this make sense?

**Comment by jeddy3** [comment]
> Does this make sense?

Absolutely. 

I think we're on the same page and I think I went through the same thought process as you (both in terms of only being able to compare properties within declaration blocks, and not inversing the principle) :)

I used `position: static` in the example above as it's aligned with the "dont-occur-together principle" i.e. `top`, `bottom`, `left` and `right` [will be ignored](https://developer.mozilla.org/en/docs/Web/CSS/position#Values) if `position` is set to `static`. Just as `float` will be ignored if `display` is set to `inline-block`.

Does that sound right?

**Comment by davidtheclark** [comment]
Oh geez. I must have just read too fast and thought you wrote position: absolute instead of static. Yeah, we're on the same page :)

**Comment by davidtheclark** [comment]
Also: I do like the name declaration-block-no-ignored-properties.

**Comment by jeddy3** [comment]
Good stuff :)

I'm marking this as "help wanted". I think it'll be pretty simple to implement the logic, but I think there's a great opportunity, if it takes someone's fancy, to go through the specs to create a definitive list of ignored properties.

**Comment by evilebottnawi** [comment]
@jeddy3 @davidtheclark What do you think about the customization of rules? We give the user the ability to use or prohibition on the use by setting an object or all warnings should lie inside the rule and not be settable?

**Comment by davidtheclark** [comment]
@evilebottnawi I'm afraid I don't understand. Could you try to explain a little more?

**Comment by evilebottnawi** [comment]
@davidtheclark see https://github.com/stylelint/stylelint/pull/994. We should array of properties inside rule or settable by user?

**Comment by jeddy3** [comment]
> should array of properties inside rule

I think inside the rule as you have it looks great. A user should be able to just turn on this rule and have it work as expected. There's no need to make it configurable.

**Comment by evilebottnawi** [comment]
@jeddy3 @davidtheclark i created stated list of ignored properties, but i could miss out on some property. If we helped get all the properties together here it would be awesome.

**Comment by evilebottnawi** [comment]
@jeddy3 also wip

**Comment by jeddy3** [comment]
>  i created stated list of ignored properties, 

For anyone who is wondering, the list is here:
https://github.com/evilebottnawi/stylelint/blob/declaration-block-no-ignored-properties/src/rules/declaration-block-no-ignored-properties/index.js#L15-L54

@evilebottnawi It's looking great so far. How did you come up with this original list? Did you find a dedicated resource about it somewhere or did you go looking through the specs?

Off hand, I can only think of `position: static` (ref: https://developer.mozilla.org/en/docs/Web/CSS/top) e.g.

```js 
{
  property: "position",
  value: "static",
  ignoredProperties: [
    "bottom",
    "left",
    "right",
    "top",
  ],
}
```

Does anyone know of an resources about this? Otherwise we either need to do an audit of the specs, or we can release this as-is and then add additional items to the array as users report them?

**Comment by evilebottnawi** [comment]
@jeddy3 see spec and based on rule https://github.com/CSSLint/csslint/wiki/Require-properties-appropriate-for-display (aso see issue about this rule). About `static` i get from css spec:
https://www.w3.org/TR/CSS2/visuren.html#propdef-position
```
static
The box is a normal box, laid out according to the normal flow. The 'top', 'right', 'bottom', and 'left' properties do not apply.
```

**Comment by evilebottnawi** [comment]
@jeddy3 I'll try collect the full list as I can (in PR we can discuss my list), other items adding as users report them is best solution.

**Comment by evilebottnawi** [comment]
@jeddy3 close?

**Comment by jeddy3** [comment]
Yep. Thanks again for this rule. It's a really good one!
